CARDS-2577: New answer type: identifier

acrowthe opened the pull request:
- Add new identifier type question
  - Currently supports auto-generated UUIDs
  - Optional "Copy to clipboard" button in both edit and view mode
- Adds new test form (Identifier Question Test"

acrowthe committed:
CARDS-2577: New answer type: identifier
- Add new identifier type question
  - Currently supports auto-generated UUIDs
  - Optional "Copy to clipboard" button in both edit and view mode
- Adds new test form (Identifier Question Test"

veronikaslc commented on modules/data-entry/src/main/frontend/src/questionnaire/IdentifierQuestion.jsx:
```diff
@@ -0,0 +1,117 @@
+  }).isRequired
+};
+
+IdentifierQuestion.defaultProps = {
```
Lets adopt new React-19-friendly way of setting defaults

marta- commented on modules/data-entry/src/main/frontend/src/questionnaireEditor/Question.json:
```diff
@@ -493,6 +493,15 @@
         "strokeColor": "string",
         "overrideMapColors": "boolean"
+      },
+      "identifier": {
```
Can we use a Mui layout component instead of custom styling? If not, can this styling be moved in the identifier question jsx, as we're trying to slowly move away from the QuestionnaireStyle monolith?

acrowthe committed:
CARDS-2577: New answer type: identifier
Add questionnaire importer support

acrowthe commented:
![image](https://github.com/user-attachments/assets/734e3c51-28b8-4c92-acca-77cd1a21187e)
Rebased and importer support added, comments have not been addressed yet

sdumitriu commented on modules/data-entry/src/main/frontend/src/questionnaire/IdentifierQuestion.jsx:
```diff
@@ -0,0 +1,117 @@
+//             Defaults to "plain".
+//
+// Sample usage:
+// <NumberQuestion
```
Why `NumberQuestion`?

sdumitriu commented on modules/data-entry/src/main/frontend/src/questionnaire/IdentifierQuestion.jsx:
```diff
@@ -0,0 +1,117 @@
+// <NumberQuestion
+//    text="Visit Identifier"
+//    displayMode="plain+copy"
+//    dataType="identifier"
```
Is this needed? Or should it instead be nested in `questionDefinition`?

sdumitriu commented on modules/data-model/forms/api/src/main/java/io/uhndata/cards/forms/api/FormUtils.java:
```diff
@@ -59,6 +59,9 @@ public interface FormUtils
     String ANSWER_SECTION_RESOURCE = "cards/AnswerSection";
 
+    /** The Sling resource super type of an AnswerSection. */
+    String ANSWER_SECTION_SUPERTYPE = "cards/Resource";
```
```suggestion
    String ANSWER_SECTION_SUPERTYPE = "cards/ResourcePart";
```

sdumitriu commented on modules/data-model/forms/api/src/main/resources/SLING-INF/nodetypes/forms.cnd:
```diff
@@ -747,6 +754,17 @@
 //-----------------------------------------------------------------------------
+// Identifier answers, stored as strings
+[cards:IdentifierAnswer] > cards:Answer
+  - value (string)
```
Shouldn't this be protected?

acrowthe replied:
Per discussion, can't make this protected as front end needs to be able to generate due to recurrent sections

sdumitriu commented on modules/data-model/forms/impl/src/main/java/io/uhndata/cards/forms/internal/FormUpdateUtilsImpl.java:
```diff
@@ -0,0 +1,151 @@
+    {
+        final String uuid = UUID.randomUUID().toString();
+        NodeBuilder node = parent.setChildNode(uuid);
+        SimpleDateFormat dateFormat = new SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss.SSSXXX");
```
We can use `DateUtils.toString(Calendar)`.

sdumitriu commented on modules/data-model/forms/impl/src/main/java/io/uhndata/cards/forms/internal/IdentifierAnswerEditorProvider.java:
```diff
@@ -0,0 +1,68 @@
+    {
+        final ResourceResolver resolver = this.rrp.getThreadResourceResolver();
+        if (resolver != null) {
+            // Each QuestionMatrixEditor maintains a state, so a new instance must be returned each time
```
```suggestion
            // Each IdentifierAnswerEditor maintains a state, so a new instance must be returned each time
```

sdumitriu commented on modules/data-model/forms/impl/src/main/java/io/uhndata/cards/forms/internal/IdentifierAnswerEditor.java:
```diff
@@ -0,0 +1,171 @@
+import io.uhndata.cards.forms.api.QuestionnaireUtils;
+
+/**
+ * An {@link Editor} that fills out any reference answers for a new form.
```
```suggestion
 * An {@link Editor} that fills out any identifier answers for a new form.
```

sdumitriu commented on modules/data-model/forms/impl/src/main/java/io/uhndata/cards/forms/internal/IdentifierAnswerEditor.java:
```diff
@@ -0,0 +1,171 @@
+
+    /**
+     * Simple constructor.
+     *ms-appid:W~com.squirrel.slack.slack
```
```suggestion
```

sdumitriu reviewed (changes requested):
An IdentifierAnswer is still an editable answer in the backend, it's just that the frontend makes is appear read-only. Do we ever want to allow identifier answers to be editable? If not, I think it would be better to do these changes:

- make the `value` property of the `cards:IdentifierAnswer` `autocreated protected`
- move the support for the `+copy` display mode into `AutocreatedAnswer`
- replace the `IdentifierQuestion.jsx` code with something like `ReferenceQuestion.jsx` but registered when `dataType==identifier`

sdumitriu commented:
And another change that I think would be good, but it can wait for later, is to not have all possible identifier types handled by the same java class, but each identifier type is implemented by its own subservice.

marta- commented on test-resources/src/main/resources/SLING-INF/content/Questionnaires/IdentifierTest.xml:
```diff
@@ -0,0 +1,88 @@
+	<primaryNodeType>cards:Questionnaire</primaryNodeType>
+	<property>
+		<name>title</name>
+		<value>Identifier Question Test</value>
```
Update the test form with an identifier question inside a section.

marta- reviewed (changes requested):
Blocker bug: the answers of identifier questions that reside inside a section don't appear to be populated.

marta- commented on modules/data-entry/src/main/frontend/src/questionnaire/IdentifierQuestion.jsx:
```diff
@@ -0,0 +1,117 @@
+      preventDefaultView
+      {...props}
+      >
+        { pageActive && <>
+          <FormattedText className={classes.identifierQuestionText}>{initialValue}</FormattedText>
+          { displayMode.endsWith("+copy") &&
+            <Tooltip title={text} onClose={handleClose} className={classes.identifierQuestionButton}>
+              <IconButton onClick={handleClick}>
+                <ContentCopyIcon />
+              </IconButton>
+            </Tooltip>
+          }
+          <div style={{clear: "both"}}></div>
+        </> }
```
```suggestion
        { pageActive &&
          ( displayMode.endsWith("+copy") ?
            <Tooltip title={text} onClose={handleClose}>
              <Button
                onClick={handleClick}
                endIcon={<ContentCopyIcon />}
                sx={{padding: 0, textTransform: "none"}}
              >
              { initialValue}
              </Button>
            </Tooltip>
            : <FormattedText on>{ initialValue }</FormattedText>
          )
```

I like how it stands out thanks to the primary color, and was wondering if it would be a good idea to display it as a button and have the tooltip and the copy functionality enabled even when the `displayMode` is `plain` (and simply not show the `Copy` icon in that case).

![image](https://github.com/user-attachments/assets/e3f4c837-c359-454c-9b32-1d5460af98da)

marta- replied:
Another look option:

![image](https://github.com/user-attachments/assets/0a4a6251-f362-41af-9fd7-67760bfc6054)

done with

```
                <FormattedText sx={{ color: theme => theme.palette.text.primary }}>
                  {"```" + initialValue + "```"}
                </FormattedText>
```

acrowthe replied:
![image](https://github.com/user-attachments/assets/f3f1feaf-5dbc-4c80-a695-a0e3c887fc75)

marta- commented:
Is this log entry something we should be looking into? @acrowthe @sdumitriu 

```
05.04.2025 20:06:09.079 *WARN* [oak-executor-321] org.apache.jackrabbit.oak.plugins.index.Cursors$TraversingCursor Traversed 17000 nodes with filter Filter(query=SELECT a.* FROM [cards:IdentifierAnswer] AS a WHERE a.copiedFrom = '/Forms/0030c2a0-734a-4b28-895d-00487b8bfeda/0399cf9a-4bee-46f8-beaa-84ff01d901c8' UNION SELECT a.* FROM [cards:IdentifierAnswer] AS a  INNER JOIN [cards:Form] AS f ON a.form = f.[jcr:uuid]  INNER JOIN [cards:Question] AS q ON a.question = q.[jcr:uuid]  WHERE    a.value is null    AND q.question = '/Questionnaires/ad1623ce-66fe-4613-8b34-3d9cfe667ceb/id3'    AND f.relatedSubjects = 'e1e384a0-c7eb-4905-9469-a8b185efbe46' OPTION (index tag cards), path=*, property=[:indexTag=[cards], form=[3c4535f8-b004-47b0-8093-7d65c18ed6a1, is not null], question=[is not null], value=[is null]]) called by (<function not configured>); consider creating an index or changing the query
```

sdumitriu commented:
Ah, we need to add an indexing entry.

On Sat, Apr 5, 2025, 20:47 Marta Girdea ***@***.***> wrote:

> Is this log entry something we should be looking into? @acrowthe
> <https://github.com/acrowthe> @sdumitriu <https://github.com/sdumitriu>
>
> 05.04.2025 20:06:09.079 *WARN* [oak-executor-321] org.apache.jackrabbit.oak.plugins.index.Cursors$TraversingCursor Traversed 17000 nodes with filter Filter(query=SELECT a.* FROM [cards:IdentifierAnswer] AS a WHERE a.copiedFrom = '/Forms/0030c2a0-734a-4b28-895d-00487b8bfeda/0399cf9a-4bee-46f8-beaa-84ff01d901c8' UNION SELECT a.* FROM [cards:IdentifierAnswer] AS a  INNER JOIN [cards:Form] AS f ON a.form = f.[jcr:uuid]  INNER JOIN [cards:Question] AS q ON a.question = q.[jcr:uuid]  WHERE    a.value is null    AND q.question = '/Questionnaires/ad1623ce-66fe-4613-8b34-3d9cfe667ceb/id3'    AND f.relatedSubjects = 'e1e384a0-c7eb-4905-9469-a8b185efbe46' OPTION (index tag cards), path=*, property=[:indexTag=[cards], form=[3c4535f8-b004-47b0-8093-7d65c18ed6a1, is not null], question=[is not null], value=[is null]]) called by (<function not configured>); consider creating an index or changing the query
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/data-team-uhn/cards/pull/1821#issuecomment-2781155120>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAAVUV7Y2L3OPAPXNMCMKBT2YB2RJAVCNFSM6AAAAABQKMBDX6VHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDOOBRGE2TKMJSGA>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
> [image: marta-]*marta-* left a comment (data-team-uhn/cards#1821)
> <https://github.com/data-team-uhn/cards/pull/1821#issuecomment-2781155120>
>
> Is this log entry something we should be looking into? @acrowthe
> <https://github.com/acrowthe> @sdumitriu <https://github.com/sdumitriu>
>
> 05.04.2025 20:06:09.079 *WARN* [oak-executor-321] org.apache.jackrabbit.oak.plugins.index.Cursors$TraversingCursor Traversed 17000 nodes with filter Filter(query=SELECT a.* FROM [cards:IdentifierAnswer] AS a WHERE a.copiedFrom = '/Forms/0030c2a0-734a-4b28-895d-00487b8bfeda/0399cf9a-4bee-46f8-beaa-84ff01d901c8' UNION SELECT a.* FROM [cards:IdentifierAnswer] AS a  INNER JOIN [cards:Form] AS f ON a.form = f.[jcr:uuid]  INNER JOIN [cards:Question] AS q ON a.question = q.[jcr:uuid]  WHERE    a.value is null    AND q.question = '/Questionnaires/ad1623ce-66fe-4613-8b34-3d9cfe667ceb/id3'    AND f.relatedSubjects = 'e1e384a0-c7eb-4905-9469-a8b185efbe46' OPTION (index tag cards), path=*, property=[:indexTag=[cards], form=[3c4535f8-b004-47b0-8093-7d65c18ed6a1, is not null], question=[is not null], value=[is null]]) called by (<function not configured>); consider creating an index or changing the query
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/data-team-uhn/cards/pull/1821#issuecomment-2781155120>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAAVUV7Y2L3OPAPXNMCMKBT2YB2RJAVCNFSM6AAAAABQKMBDX6VHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDOOBRGE2TKMJSGA>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

acrowthe committed:
CARDS-2577: New answer type: identifier
- Cleanup code and address review comments
- Add frontend support for generating identifiers
  - Required for identifiers in recurrent sections

marta- commented on modules/data-entry/src/main/frontend/src/questionnaire/IdentifierQuestion.jsx:
```diff
@@ -0,0 +1,118 @@
+    displayMode = "plain",
+    identifierType = "uuid"
+  } = {...props.questionDefinition };
+  const [ text, setText ] = useState("Copy to Clipboard")
```
```suggestion
  const [ text, setText ] = useState("Copy to clipboard")
```

marta- commented on modules/data-entry/src/main/frontend/src/questionnaire/IdentifierQuestion.jsx:
```diff
@@ -0,0 +1,118 @@
+  const [ text, setText ] = useState("Copy to Clipboard")
+
+  const [value, setValue] = useState(existingAnswer?.[1]?.value || "");
+  const answer = [["value", value]];
```
What is the meaning of the first position of the inner array ("value")?

acrowthe replied:
The `[1]` is the answer node.
Existing value has `[0]` as the nodes key/name (a UUID) and then `[1]` as the actual node object. As a result the magic `existingAnswer?.[1]` is quite common in question.jsx files

marta- replied:
After a quick look at MultipleChoice, it would appear that position [0] of the answer passed to the <Answer> component might generally be the label. If that's the case, it might make more sense to pass this instead:

```suggestion
  const answer = [[value, value]];
```

marta- commented on modules/data-entry/src/main/frontend/src/questionnaire/IdentifierQuestion.jsx:
```diff
@@ -0,0 +1,118 @@
+  }
+
+  const handleClose = () => {
+    setText("Copy to Clipboard");
```
Can we define "Copy to clipboard" as a constant?

marta- commented on modules/data-model/forms/api/src/main/resources/SLING-INF/nodetypes/forms.cnd:
```diff
@@ -747,6 +754,17 @@
+  - value (string)
+
+  // Hardcode the resource type.
+  - sling:resourceType (STRING) = "cards/SelectableAreaAnswer" mandatory autocreated protected
```
Copy/paste artifact

acrowthe committed:
CARDS-2577: New answer type: identifier
Fix backend identifier generation to work with sections
Minor front end code improvements

sdumitriu committed:
CARDS-2577: New answer type: identifier

Better usage of `answers[LABEL_POS]`

Co-authored-by: Marta Girdea <651980+marta-@users.noreply.github.com>

sdumitriu commented on modules/data-entry/src/main/frontend/src/questionnaire/IdentifierQuestion.jsx:
```diff
@@ -0,0 +1,117 @@
+// Component that renders an identifier question, with optional copy button.
+//
+// Optional props:
+//  text: String containing text to show alongside the identifier
```
I don't think this belongs here. `text` is a standard Question prop, and it is taken from `questionDefinition`, the question's text.

sdumitriu committed:
CARDS-2577: New answer type: identifier
Better documentation for the React component

sdumitriu committed:
CARDS-2577: New answer type: identifier
Add @since to the new java class
